src/metabase/db/connection_pool_setup.clj

```diff
@@ -1,17 +1,45 @@
 (ns metabase.db.connection-pool-setup
   "Code for creating the connection pool for the application DB and setting it as the default Toucan connection."
-  (:require [metabase.config :as config]
+  (:require [java-time :as t]
+            [metabase.config :as config]
             [metabase.connection-pool :as connection-pool]
             [schema.core :as s])
-  (:import com.mchange.v2.c3p0.PoolBackedDataSource))
+  (:import [com.mchange.v2.c3p0 ConnectionCustomizer PoolBackedDataSource]))
+
+(def ^:private latest-checkin (atom nil))
+
+(def ^:private recent-window-seconds 15)
+
+(defn recent-activity?
+  "Returns true if there has been recent activity. Define recent activity as an application db connection checked in
+  within 15 seconds. Check-in means a query succeeded and the db connection is no longer needed."
+  []
+  (when-let [activity @latest-checkin]
+    (t/after? activity (t/minus (t/offset-date-time) (t/seconds recent-window-seconds)))))
```

Comment:
We should write a test for this

- `with-redefs` the `recent-window` to something super short like 1 millisecond
   - Thread sleep for 2 milliseconds just to be extra safe
   - `recent-activity?` should return `false`
- (outside of the `with-redefs`)
   - check out a connection from the app DB (`jdbc/with-db-connection [_conn (db/connection)]` or something would work 
   - `recent-activity?` should return `true` now